relay-general/src/protocol/thread.rs

```diff
@@ -69,6 +71,116 @@ impl Empty for ThreadId {
     }
 }
 
+#[derive(Debug, Copy, Clone, Eq, PartialEq, ProcessValue, Empty)]
+#[cfg_attr(feature = "jsonschema", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "jsonschema", schemars(rename_all = "lowercase"))]
+pub enum LockReasonType {
+    Locked = 1,
+    Waiting = 2,
+    Sleeping = 4,
+    Blocked = 8,
+    // This enum does not have a `fallback_variant` because we consider it unlikely to be extended. If it is,
+    // The error added to `Meta` will tell us to update this enum.
+}
+
+impl LockReasonType {
+    fn from_android_lock_reason_type(value: u64) -> Option<LockReasonType> {
+        Some(match value {
+            1 => LockReasonType::Locked,
+            2 => LockReasonType::Waiting,
+            4 => LockReasonType::Sleeping,
+            8 => LockReasonType::Blocked,
+            _ => return None,
+        })
+    }
+}
+
+impl fmt::Display for LockReasonType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            LockReasonType::Locked => write!(f, "locked"),
+            LockReasonType::Waiting => write!(f, "waiting"),
+            LockReasonType::Sleeping => write!(f, "sleeping"),
+            LockReasonType::Blocked => write!(f, "blocked"),
+        }
+    }
+}
+
+impl FromValue for LockReasonType {
+    fn from_value(value: Annotated<Value>) -> Annotated<Self> {
+        match value {
+            Annotated(Some(Value::U64(val)), mut meta) => {
+                match LockReasonType::from_android_lock_reason_type(val) {
+                    Some(value) => Annotated(Some(value), meta),
+                    None => {
+                        meta.add_error(ErrorKind::InvalidData);
+                        meta.set_original_value(Some(val));
+                        Annotated(None, meta)
+                    }
+                }
+            }
+            Annotated(Some(Value::I64(val)), mut meta) => {
+                match LockReasonType::from_android_lock_reason_type(val as u64) {
+                    Some(value) => Annotated(Some(value), meta),
+                    None => {
+                        meta.add_error(ErrorKind::InvalidData);
+                        meta.set_original_value(Some(val));
+                        Annotated(None, meta)
+                    }
+                }
+            }
+            Annotated(None, meta) => Annotated(None, meta),
+            Annotated(Some(value), mut meta) => {
+                meta.add_error(Error::expected("lock reason type"));
+                meta.set_original_value(Some(value));
+                Annotated(None, meta)
+            }
+        }
+    }
+}
+
+impl IntoValue for LockReasonType {
+    fn into_value(self) -> Value {
+        Value::String(self.to_string())
```

Comment:
I missed this on my previous review: An upstream Relay has to be able to parse this again after re-serialization, so we have to serialize as an integer (see related suggestions below).

```suggestion
        Value::U64(self as u64)
```

Reply:
Ah I see 👍